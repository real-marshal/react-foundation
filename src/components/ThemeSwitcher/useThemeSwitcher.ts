import { useEffect, useState } from 'react'
import { useMediaQuery } from '../../hooks/useMediaQuery'

const defaultTheme = 'light'

export const useThemeSwitcher = () => {
  const prefersDark = useMediaQuery('(prefers-color-scheme: dark)')
  const prefersLight = useMediaQuery('(prefers-color-scheme: light)')

  const persistedTheme = localStorage.getItem('theme')

  const [theme, setTheme] = useState(persistedTheme ?? 'system')

  useEffect(() => {
    const actualTheme =
      theme === 'system'
        ? // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
          (prefersDark && 'dark') || (prefersLight && 'light') || defaultTheme
        : theme

    document.body.dataset['theme'] = actualTheme

    localStorage.setItem('theme', theme)
  }, [theme, prefersDark, prefersLight])

  return { theme, setTheme }
}
