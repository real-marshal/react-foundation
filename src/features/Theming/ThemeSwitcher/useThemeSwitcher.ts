import { useEffect, useState } from 'react'
import { useMediaQuery } from '@/hooks'

const defaultThemeFallback = 'light'

export const useThemeSwitcher = (defaultTheme = 'system') => {
  const prefersDark = useMediaQuery('(prefers-color-scheme: dark)')
  const prefersLight = useMediaQuery('(prefers-color-scheme: light)')

  const persistedTheme = localStorage.getItem('theme')

  const [theme, setTheme] = useState(persistedTheme ?? defaultTheme)

  useEffect(() => {
    const actualTheme =
      theme === 'system'
        ? // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
          (prefersDark && 'dark') || (prefersLight && 'light') || defaultThemeFallback
        : theme

    document.body.dataset['theme'] = actualTheme

    localStorage.setItem('theme', theme)
  }, [theme, prefersDark, prefersLight])

  return { theme, setTheme }
}
