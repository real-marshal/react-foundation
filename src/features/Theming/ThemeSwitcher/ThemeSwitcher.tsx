import { useThemeSwitcher } from './useThemeSwitcher'
import { capitalize } from '@/utils'
import { List } from '@/components'

const themes = ['system', 'light', 'dark'] as const

const ThemeSwitcher = () => {
  const { theme: currentTheme, setTheme } = useThemeSwitcher()

  return (
    <List>
      {themes.map((theme) => (
        <List.ListItem
          key={theme}
          selected={currentTheme === theme}
          onClick={() => setTheme(theme)}
          value={theme}
        >
          {capitalize(theme)}
        </List.ListItem>
      ))}
    </List>
  )
}

export { ThemeSwitcher }
