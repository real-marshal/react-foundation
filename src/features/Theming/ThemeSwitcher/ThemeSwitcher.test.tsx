import { composeStories } from '@storybook/react'
import { render } from '@testing-library/react'
import * as stories from './ThemeSwitcher.stories'

jest.mock('./useThemeSwitcher', () => ({
  useThemeSwitcher: () => ({ theme: 'system', setTheme: () => null }),
}))

describe('ThemeSwitcher component', () => {
  const testStories = Object.values(composeStories(stories)).map(
    (Story) => [Story.storyName, Story] as const
  )

  it.each(testStories)('renders %s story', (_, Story) => {
    const view = render(<Story />)
    expect(view.baseElement).toMatchSnapshot()
  })
})
