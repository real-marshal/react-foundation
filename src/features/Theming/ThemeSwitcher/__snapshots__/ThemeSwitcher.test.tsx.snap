// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSwitcher component renders Basic story 1`] = `
.emotion-0 {
  border: 1px solid black;
  border-radius: 10px;
  padding: 2px;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

.emotion-2 {
  list-style: none;
  border: 1px solid #333;
  border-radius: 5px;
  padding: 4px 12px;
  margin: 2px;
  cursor: pointer;
  background-color: #aaa;
}

.emotion-2:hover {
  background-color: #ccc;
}

.emotion-4 {
  list-style: none;
  border: 1px solid #333;
  border-radius: 5px;
  padding: 4px 12px;
  margin: 2px;
  cursor: pointer;
}

.emotion-4:hover {
  background-color: #ccc;
}

<body>
  <div>
    <ul
      class="emotion-0 emotion-1"
    >
      <li
        class="emotion-2 emotion-3"
        value="system"
      >
        System
      </li>
      <li
        class="emotion-4 emotion-3"
        value="light"
      >
        Light
      </li>
      <li
        class="emotion-4 emotion-3"
        value="dark"
      >
        Dark
      </li>
    </ul>
  </div>
</body>
`;
